ppi.py

```diff
@@ -0,0 +1,89 @@
+                            if "DESCRIPTOR" in type_name:
+                                target_param = params.index(param)
+                                break
+                        if target_param is None:
+                            continue
+
+                        param = instr.params[target_param]
+                        if not isinstance(param, MediumLevelILConstPtr):
+                            continue
+
+                        var_addr = param.constant
+                        sym = bv.get_symbol_at(var_addr)
+
+                        if sym is not None:
+                            var_name = sym.name
+                        else:
+                            var_name = None
+
+                        if type_name[-1] == '*':
+                            type_name = type_name[:-1]
+
+                        bv.define_user_data_var(var_addr, type_name, var_name)
+                        bv.update_analysis_and_wait()
+
+                        notify_descriptor = bv.get_data_var_at(var_addr).value
+
+                        if not isinstance(notify_descriptor, dict):
+                            continue
+                        if 'Guid' not in notify_descriptor.keys():
+                            continue
+
+                        # define types for guid and notify entrypoint
+                        guid = bv.read(notify_descriptor['Guid'], 16)
+                        if (not guid) or len(guid) != 16:
+                            continue
+                        guid_name, protocol_name = lookup_protocol_guid(guid)
+                        bv.define_user_data_var(notify_descriptor['Guid'], 'EFI_GUID', guid_name)
+
+                        if 'Notify' in notify_descriptor.keys():
+                            notify_entrypoint = notify_descriptor['Notify']
+                            func = bv.get_function_at(notify_entrypoint)
+                            if not func:
+                                continue
+                            if protocol_name:
+                                func_name = f"Notify_{protocol_name}"
+                            else:
+                                func_name = f"Notify_UNKNOWN_PPI"
+                            func.type = f"EFI_STATUS {func_name}(EFI_PEI_SERVICES **PeiServices, EFI_PEI_NOTIFY_DESCRIPTOR* NotifyDescriptor, VOID* Ppi)"
+                            bv.update_analysis_and_wait()
+                            if not propagate_system_table_pointers(bv, task, func):
+                                return False
+
+                        elif 'Ppi' in notify_descriptor.keys():
+                            ppi_addr = notify_descriptor['Ppi']
+                            if not protocol_name:
+                                ppi_name = 'UNKNOWN_PPI'
+                            else:
+                                ppi_name = protocol_name
+                            bv.define_user_data_var(ppi_addr, 'VOID*', ppi_name)
```

Comment:
Don't we have a few `PPI` protocol types defined in our EFI platform types? We should add all of the PPI protocols from EDK-II and assign them, instead of just making everything a `void *`

Reply:
I have already added all PPIs from EDK2, but it's very likely to encounter third-party PPIs in firmware. I wonder if it is possible to leverage the structure auto-fields feature to create custom structures for unknown PPIs.